File: deno/src/cron/job/job.dao.ts
import {
  validateOption as validateOptionJob,
  validateIsEnabled as validateIsEnabledJob,
  findById as findByIdJob,
} from "/gen/cron/job/job.dao.ts";

import {
  create as createCronJobLog,
  updateById as updateByIdCronJobLog,
} from "/gen/cron/cron_job_log/cron_job_log.dao.ts";

import {
  create as createCronJobLogDetail,
} from "/gen/cron/cron_job_log_detail/cron_job_log_detail.dao.ts";

import dayjs from "dayjs";

import {
  CronJobLogExecState,
} from "/gen/types.ts";

export async function runJob(
  job_id: JobId,
  cron_job_id: CronJobId,
  cron: string,
  tenant_id: TenantId,
) {
  const begin_time = dayjs().format("YYYY-MM-DD HH:mm:ss");
  const id = await createCronJobLog({
    cron_job_id,
    begin_time,
    exec_result: "",
    tenant_id,
  });
  let exec_result: string | undefined | void;
  try {
    const job_model = await validateOptionJob(
      await findByIdJob(job_id),
    );
    await validateIsEnabledJob(job_model);
    console.log(`cron_job begin: ${ job_model.code }: ${ cron }`);
    const code = job_model.code;
    if (code === "test") {
      exec_result = await test(id, tenant_id);
    }
    exec_result = exec_result || "";
    const end_time = dayjs().format("YYYY-MM-DD HH:mm:ss");
    await updateByIdCronJobLog(
      id,
      {
        exec_state: CronJobLogExecState.Success,
        exec_result,
        end_time,
      },
    );
  } catch (err0) {
    const err = err0 as Error;
    const end_time = dayjs().format("YYYY-MM-DD HH:mm:ss");
    const exec_result = (err.message || err.toString() || "").substring(0, 4000);
    await updateByIdCronJobLog(
      id,
      {
        exec_state: CronJobLogExecState.Fail,
        exec_result,
        end_time,
      },
    );
    throw err;
  } finally {
    console.log(`cron_job end: ${ job_id }: ${ cron }`);
  }
  return exec_result || "";
}

async function test(
  cron_job_log_id: CronJobLogId,
  tenant_id: TenantId,
) {
  await createCronJobLogDetail({
    lbl: "测试日志",
    cron_job_log_id,
    tenant_id,
  });
  await new Promise((resolve) => setTimeout(resolve, 1000));
  console.log("test");
}

File: rust/src/src/cron/job/job_dao.rs
use color_eyre::eyre::{Result, eyre};
use tracing::info;

use crate::common::context::Options;

use crate::r#gen::base::tenant::tenant_model::TenantId;
use crate::r#gen::cron::job::job_model::JobId;
use crate::r#gen::cron::cron_job::cron_job_model::CronJobId;
use crate::r#gen::cron::cron_job_log::cron_job_log_model::{
  CronJobLogId,
  CronJobLogInput,
  CronJobLogExecState,
};
use crate::r#gen::cron::cron_job_log_detail::cron_job_log_detail_model::CronJobLogDetailInput;

use crate::r#gen::cron::job::job_dao::{
  validate_option_job,
  validate_is_enabled_job,
  find_by_id_job,
};

use crate::r#gen::cron::cron_job_log::cron_job_log_dao::{
  create_cron_job_log,
  update_by_id_cron_job_log,
};

use crate::r#gen::cron::cron_job_log_detail::cron_job_log_detail_dao::create_cron_job_log_detail;

pub async fn run_job(
  id: JobId,
  cron_job_id: CronJobId,
  cron: String,
  tenant_id: TenantId,
  options: Option<Options>,
) -> Result<String> {
  
  let job_model = validate_option_job(
    find_by_id_job(
      id,
      options.clone(),
    ).await?,
  ).await?;
  
  validate_is_enabled_job(
    &job_model,
  ).await?;
  
  info!("cron_job begin: {}: {}", cron_job_id, cron);
  
  let begin_time = chrono::Local::now().naive_local();
  let id = create_cron_job_log(
    CronJobLogInput {
      cron_job_id: Some(cron_job_id.clone()),
      begin_time: Some(begin_time),
      exec_result: Some(String::new()),
      tenant_id: Some(tenant_id.clone()),
      ..Default::default()
    },
    options.clone(),
  ).await?;
  
  let code = job_model.code;
  
  let exec_result: Option<Result<String>> = if code == "test" {
    test(
      id.clone(),
      tenant_id,
    ).await.into()
  } else {
    None
  };
  
  info!("cron_job end: {}: {}", cron_job_id, cron);
  
  let end_time = chrono::Local::now().naive_local();
  
  if exec_result.is_some() && exec_result.as_ref().unwrap().is_err() {
    let exec_result = exec_result.unwrap();
    let exec_result = exec_result.unwrap_err().to_string();
    update_by_id_cron_job_log(
      id,
      CronJobLogInput {
        exec_state: Some(CronJobLogExecState::Fail),
        exec_result: Some(exec_result.clone()),
        end_time: Some(end_time),
        ..Default::default()
      },
      options.clone(),
    ).await?;
    
    return Err(eyre!(exec_result));
  }
  
  let exec_result: String = {
    if let Some(exec_result) = exec_result {
      exec_result.unwrap()
    } else {
      String::new()
    }
  };
  
  update_by_id_cron_job_log(
    id,
    CronJobLogInput {
      exec_state: Some(CronJobLogExecState::Success),
      exec_result: Some(exec_result.clone()),
      end_time: Some(end_time),
      ..Default::default()
    },
    options.clone(),
  ).await?;
  
  Ok(exec_result)
}

async fn test(
  cron_job_log_id: CronJobLogId,
  tenant_id: TenantId,
) ->Result<String> {
  
  create_cron_job_log_detail(
    CronJobLogDetailInput {
      lbl: Some("测试日志".to_string()),
      cron_job_log_id: Some(cron_job_log_id),
      tenant_id: Some(tenant_id),
      ..Default::default()
    },
    None,
  ).await?;
  
  tokio::time::sleep(std::time::Duration::from_secs(3)).await;
  
  info!("test");
  
  Ok(String::new())
}
